hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/ozone/container/common/states/datanode/RunningDatanodeState.java

```diff
@@ -0,0 +1,292 @@
+   *
+   * @param idPath - Path to the ID File.
+   * @return DatanodeID
+   * @throws IOException
+   */
+  private StorageContainerDatanodeProtocolProtos.ContainerNodeIDProto
+      readPersistedDatanodeID(Path idPath) throws IOException {
+    Preconditions.checkNotNull(idPath);
+    StorageContainerDatanodeProtocolProtos.ContainerNodeIDProto
+        containerIDProto;
+    try (FileInputStream stream = new FileInputStream(idPath.toFile())) {
+      containerIDProto = StorageContainerDatanodeProtocolProtos
+          .ContainerNodeIDProto.parseFrom(stream);
+      return containerIDProto;
+    }
+  }
+
+  /**
+   * Create a DatanodeID from the datanode information.
+   *
+   * @return DatanodeID
+   * @throws UnknownHostException
+   */
+  private DatanodeID createDatanodeID() throws UnknownHostException {
+    DatanodeID temp = new DatanodeID(
+        //TODO : Replace this with proper network and kerberos
+        // support code.
+        InetAddress.getLocalHost().getHostAddress().toString(),
+        DataNode.getHostName(conf),
+        UUID.randomUUID().toString(),
+        0, /** XferPort - SCM does not use this port  */
+        0, /** Info port - SCM does not use this port */
+        0, /** Info Secure Port - SCM does not use this port */
+        0); /** IPC port - SCM does not use this port */
+
+    // TODO: make this dynamically discoverable. SCM can hand out this
+    // port number to calling applications. This makes it easy to run multiple
+    // container endpoints on the same machine.
+    temp.setContainerPort(OzoneClientUtils.getContainerPort(conf));
+    return temp;
+  }
+
+  /**
+   * Creates a new ContainerID that persists both DatanodeID and ClusterID.
+   *
+   * @param idPath Path to the id file.
+   * @return ContainerNodeIDProto
+   * @throws UnknownHostException
+   */
+  private StorageContainerDatanodeProtocolProtos.ContainerNodeIDProto
+      createNewContainerID(Path idPath)
+      throws IOException {
+    StorageContainerDatanodeProtocolProtos.ContainerNodeIDProto
+        containerIDProto = StorageContainerDatanodeProtocolProtos
+        .ContainerNodeIDProto.newBuilder()
+        .setDatanodeID(createDatanodeID().getProtoBufMessage()).build();
+    try (FileOutputStream stream = new FileOutputStream(idPath.toFile())) {
+      stream.write(containerIDProto.toByteArray());
+      return containerIDProto;
+    }
+  }
+
+  /**
+   * Returns ContainerNodeIDProto or null in case of Error.
+   *
+   * @return ContainerNodeIDProto
+   */
+  private StorageContainerDatanodeProtocolProtos.ContainerNodeIDProto
+      getContainerNodeID() {
+    String dataNodeIDPath = conf.get(OzoneConfigKeys.OZONE_SCM_DATANODE_ID);
+    if (dataNodeIDPath == null || dataNodeIDPath.isEmpty()) {
+      LOG.error("A valid file path is needed for config setting {}",
+          OzoneConfigKeys.OZONE_SCM_DATANODE_ID);
+
+      // This is an unrecoverable error.
+      this.context.setState(DatanodeStateMachine.DatanodeStates.SHUTDOWN);
```

Comment:
Should we return null after mark DatanodeStates.SHUTDOWN instead of continue attempt to read the container ID?